app/src/org/commcare/logging/DataChangeLogger.kt

```diff
@@ -0,0 +1,155 @@
+package org.commcare.logging
+
+import android.content.Context
+import android.os.Environment
+import android.util.Log
+import org.apache.commons.lang3.StringUtils
+import org.commcare.android.logging.ReportingUtils
+import org.commcare.utils.CrashUtil
+import org.commcare.utils.FileUtil
+import org.javarosa.core.io.StreamsUtil
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.io.IOException
+import java.text.SimpleDateFormat
+import java.util.*
+
+/**
+ *  Used to log some crtitical platform events like DB migrations, Clear user data,
+ *  app installations etc. All these logs are backed by file storage and are accessible
+ *  from outside CommCare
+ */
+class DataChangeLogger {
+
+    fun init(context: Context) {
+        initLogFiles(context)
+    }
+
+    private fun initLogFiles(context: Context) {
+        if (!isExternalStorageWritable()) {
+            Log.e(TAG, "External Storage unavialable to write logs")
+            return
+        }
+
+        primaryFile = initFile(context, PRIMARY_LOG_FILE_NAME)
+        if (primaryFile!!.exists() && primaryFile!!.length() > MAX_FILE_SIZE) {
+            // When primary file gets full -> Delete secondary and move logs in primary to secondary
+            secondaryFile = initFile(context, SECONDARY_LOG_FILE_NAME)
+            secondaryFile!!.delete()
+            try {
+                FileUtil.copyFile(primaryFile, secondaryFile)
+                primaryFile!!.delete()
+                primaryFile!!.createNewFile()
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
+        }
+    }
+
+    private fun initFile(context: Context, fileName: String): File {
+        val file = File(context.getExternalFilesDir(null), fileName)
+        if (!file.exists()) {
+            try {
+                file.createNewFile()
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
+        }
+        return file
+    }
+
+    companion object {
+
+        private const val PRIMARY_LOG_FILE_NAME = "CommCare Data Change Logs.txt"
+        private const val SECONDARY_LOG_FILE_NAME = "CommCare Data Change Logs1.txt"
+        private const val DATE_FORMAT = "d MMM yy HH:mm:ss z"
+        private const val MAX_FILE_SIZE = (40 * 1000).toLong() // ~40KB
+
+        private val TAG = DataChangeLogger::class.java.simpleName
+
+        private var primaryFile: File? = null
+        private var secondaryFile: File? = null
+
+        /**
+         * Logs a given message to the fileSystem
+         */
+        @JvmStatic
+        fun log(message: String) {
+            // Include this info as part of any crash reports
+            CrashUtil.log(message)
+
+            // Write to local storage
+            if (primaryFile != null && primaryFile!!.exists()) {
```

Comment:
Given that you've already checked for `primaryFile != null`, what purpose does using the `!!` serve?

Reply:
It's a bit confusing for me too. But as I understand since `primaryFile` is a mutable nullable property, the kotlin compiler throws an error when I just do `primaryFile.exists` since it's possible that some other process would have changed it's value back to `null` after the null check. So adding !! is just asserting that it's still not null.

In contrast we don't need any such checks for local properties - [Example here](https://github.com/dimagi/commcare-android/blob/dataChangeLogs/app/src/org/commcare/logging/DataChangeLogger.kt#L139)

Reply:
Ah I see, so Kotlin automatically treats all function args as `val`s rather than `var`s? Do you know why that is?

Reply:
No, they both are `var`s but since Kotlin like Java follows pass-by-value semantics while calling a function, each function call has it's own local copy and it's not possible for any other process to change it.

Reply:
Haha ok that does make sense- thanks for the explanation!